service/config_provider.go

```diff
@@ -79,8 +81,13 @@ type configProvider struct {
 // * Then unmarshalls the final config.Config using the given configunmarshaler.ConfigUnmarshaler.
 //
 // Notice: This API is experimental.
-func NewConfigProvider(configMapProviders []configmapprovider.Provider, cfgMapConverters []ConfigMapConverterFunc, configUnmarshaler configunmarshaler.ConfigUnmarshaler) ConfigProvider {
+func NewConfigProvider(
+	urls []*url.URL,
+	configMapProviders map[string]configmapprovider.Provider,
```

Comment:
Is there a reason to require available providers and their scheme mappings to be included here explicitly rather than having a provider registry that providers hook into with an `init()` function?  That would seem to avoid issues that could be caused here if providers aren't included or have their scheme mapping changed around.

Reply:
That is about "initialization". I am trying at this point to focus more on the "Provider" changes to ensure I can achieve everything.

Not sure if we need the global + init() or have the "Builder" do the right thing, and have in the builder "cfgmapproviders" section. Will think more about this in a followup after I am done with `Provider` interface.

Reply:
I think it should generally be the case that the builder is a way to build a working collector, and not the only way.  To the extent that we pile responsibility for needed functionality on the builder it seems like it would become more mandatory and less optional.

Will this PR stay as a draft until the followup is completed, or at least a spec agreed?  I'd prefer to avoid churn here if there is uncertainty about the way forward.

Reply:
I understand the churn problem, but these 2 are kind of different problems: finalizing the Provider interface vs how the Providers are passed to the Collector.

I would say that manually passing the list of available providers should be the way to go:
1. Avoids using global state;
2. Consistent with other components, I would ask why are these special?
3. Global + Init() is sometimes unexpected for the distribution owners, e.g. one component that they include may by mistake or intentionally depend on a provider (call init()) and make it available in the distribution
4. The builder is not the only way. The distribution that have custom builds will have to write this map instead of just "includes", which is a bit of extra work.

Based on the first 3 reasons I would go with an API where distributions have to manually pass the available providers that they want for that distribution.

Reply:
>Consistent with other components, I would ask why are these special?

Passing a `Factories` of config providers would be consistent with what we do with other component types.

Reply:
Here we are talking about passing the available components vs register to a global registry. The name Factory vs Provider we can debate :)

Reply:
I understand. I was just supporting the idea that it does not need to be a global registry. We don't do global registries for component factories, and I don't think we need a global registry for config providers.